python/visualizer/viewer.py

```diff
@@ -45,48 +54,141 @@ def __init__(self, bb_color, x, y, width, height, label, **kwargs):
+        # This is needed for the default vertex shader.
+        self.canvas['projection_mat'] = Window.render_context['projection_mat']
+        self.canvas['modelview_mat'] = Window.render_context['modelview_mat']
+
+
 class Viewer(BoxLayout):
-    data = ObjectProperty(force_dispatch=True)
-    need_init = BooleanProperty(True)
-    dsm = ObjectProperty(None, allownone=True)
+    data = DictProperty(force_dispatch=True)
+    visualisers = ListProperty([], allownone=True)
     flipHoriz = BooleanProperty(False)
     flipVert = BooleanProperty(False)
+    settings = DictProperty({}, allownone=True)
+    title = StringProperty('Title')
     colorfmt = 'luminance'
-
-    def on_dsm(self, instance, value):
-        if self.dsm is not None:
-            self.colorfmt = self.dsm.get_colorfmt()
-            x, y = self.dsm.get_dims()
-            buf_shape = (dp(x), dp(y))
-            self.image.texture = Texture.create(size=buf_shape, colorfmt=self.colorfmt)
-            self.need_init = False
+    orientation = 'vertical'
 
     def __init__(self, **kwargs):
         super(Viewer, self).__init__(**kwargs)
-        # self.image = Image()
-        # self.add_widget(self.image)
+        self.settings_box = None
         from matplotlib.pyplot import get_cmap
         self.cm = get_cmap('tab20')
-        self.image.texture = None
-        Clock.schedule_once(self.init, 0)
-
-    def init(self, dt):
-        self.data = np.zeros((1, 1), dtype=np.uint8)
+        self.current_time = 0
+        self.current_time_window = 0
+
+    def on_visualisers(self, instance, value):
+        if self.visualisers is not None and self.visualisers:
+            for v in self.visualisers:  # TODO manage cases with multiple of below data_types
+                data_shape = v.get_dims()
+                buf_shape = (dp(data_shape[0]), dp(data_shape[1]))
+                colorfmt = v.get_colorfmt()
+                if v.data_type in ['dvs', 'frame', 'pose6q', 'point3', 'flowMap']:
+                    self.image.texture = Texture.create(size=buf_shape, colorfmt=colorfmt)
+                    self.image.canvas['image'] = 0
+                    fs_multitexture = '''
+                        $HEADER$
+
+                        uniform sampler2D image;
+                
+                        void main(void) {
+                            gl_FragColor = texture2D(image, tex_coord0);
+                        }
+                        '''
+                    self.image.canvas.shader.fs = fs_multitexture
+                elif v.data_type == 'mask':
+                    self.image.mask = Texture.create(size=buf_shape, colorfmt=colorfmt)
+                    with self.image.canvas:
+                        BindTexture(texture=self.image.mask, index=1)
+                    self.image.canvas['mask'] = 1
+                    fs_multitexture = '''
+                        $HEADER$
+                        uniform sampler2D image;
+                        uniform sampler2D mask;
```

Comment:
I'm going to need some help understanding what's going on here ... seems the only difference between frame and mask is one line in this string which gets sent to the shader - what does it do?